File: src/lib/logger.ts
/**
 * Structured logger — uses Winston for JSON output in production,
 * human-readable format in development.
 *
 * MCP servers communicate via stdio, so we log to stderr to avoid
 * contaminating the MCP protocol stream.
 */

import { createLogger, format, transports } from 'winston';

const isDev = process.env.NODE_ENV !== 'production';

export const logger = createLogger({
  level: process.env.LOG_LEVEL ?? (isDev ? 'debug' : 'info'),
  format: isDev
    ? format.combine(
        format.colorize(),
        format.timestamp({ format: 'HH:mm:ss' }),
        format.printf(({ timestamp, level, message, ...meta }) => {
          const metaStr = Object.keys(meta).length ? ` ${JSON.stringify(meta)}` : '';
          return `${timestamp} ${level}: ${message}${metaStr}`;
        })
      )
    : format.combine(format.timestamp(), format.json()),
  // CRITICAL: Log to stderr, not stdout — stdout is reserved for MCP protocol
  transports: [new transports.Stream({ stream: process.stderr })],
});

File: python-fastmcp/mcp_server/logger.py
"""Structured logging to stderr only.

CRITICAL: Never log to stdout in an MCP server.
stdout is reserved for JSON-RPC protocol messages.
Any non-JSON-RPC output on stdout corrupts the MCP session.

This module configures structlog to write to sys.stderr exclusively.
"""

import logging
import sys

import structlog

from .config import settings

# Map string log level to logging constant
_LOG_LEVELS = {
    "DEBUG": logging.DEBUG,
    "INFO": logging.INFO,
    "WARNING": logging.WARNING,
    "ERROR": logging.ERROR,
    "CRITICAL": logging.CRITICAL,
}

_level = _LOG_LEVELS.get(settings.log_level.upper(), logging.INFO)

# Configure stdlib logging — stderr only, never stdout
logging.basicConfig(
    level=_level,
    stream=sys.stderr,  # ← CRITICAL: stderr, not stdout
    format="%(message)s",
)

structlog.configure(
    processors=[
        structlog.contextvars.merge_contextvars,
        structlog.processors.add_log_level,
        structlog.processors.TimeStamper(fmt="iso"),
        structlog.dev.ConsoleRenderer(
            colors=False  # No ANSI codes — keeps stderr clean in non-TTY envs
        ),
    ],
    wrapper_class=structlog.make_filtering_bound_logger(_level),
    logger_factory=structlog.PrintLoggerFactory(file=sys.stderr),  # ← stderr
    cache_logger_on_first_use=True,
)


def get_logger(name: str) -> structlog.stdlib.BoundLogger:
    """Get a named structured logger writing to stderr."""
    return structlog.get_logger(name)
